ethsnarks/mimc.py

```diff
@@ -0,0 +1,232 @@
+"""
+https://eprint.iacr.org/2016/492.pdf
+https://eprint.iacr.org/2016/542.pdf
+https://github.com/zcash/zcash/issues/2233
+https://keccak.team/files/SpongeIndifferentiability.pdf
+
+Where the field F_p is prime, it needs to be assured that
+the cubing in the round function creates a permutation.
+For this is, it is sufficient to require gcd(n, p-1) = 1
+
+p = curve_order = 21888242871839275222246405745257275088548364400416034343698204186575808495617
+e = 5
+gcd(e, p-1) = 1
+
+The number of rounds for constructing the keyed permutation is
+
+    r = int(log(p)/log2(e)) = 75
+    r = math.ceil(log(p, e)) = 110
```

Comment:
These are much smaller than the recommendation in the MiMC paper. Section 5.1 says:

> The number of rounds for constructing the keyed permutation is r = ceiling(log2(p) / log2(3)). In the Feistel mode, we define MiMC-2p/p where the round function is defined over F<sub>p</sub> and where the number of rounds is double with respect to MiMC-p/p.

(There's a typo in the numerator of the formula but it is supposed to be log2(p); note that this is the same as ceiling(log3(field order)), in the non-Feistel case, for both binary and prime fields.)

In other words the number of rounds in the Feistel case is 2.ceiling(log2(p) / log2(3)), which for the above p is 322 rounds.

Reply:
Note that the base 3 in log<sub>3</sub>(field order) is *not* e. The reason is the analysis in section 5.3, which says:
> Using the same technique, after r rounds, the number of terms of the polynomial is upper bounded by 3<sup>r</sup> + 1, which is (much) smaller than (2<sup>t</sup> + 1)<sup>r</sup> + 1. Note that 3<sup>r</sup> + 1 is exact the same upper bounded obtained for the exponent 3 (which corresponds to t = 1). Thus, the number of rounds to guarantee the security against the algebraic attacks doesn’t change choosing exponent of the form 2<sup>t</sup> + 1 for t > 1. 

(5 is obviously of that form.)